dynamic_array.c

```diff
@@ -0,0 +1,340 @@
+// Change length of buffer
+//  * If growing, grow capacity if required and zero-fill new items
+//  * If shrinking, call destructor for any items which are being removed
+void buffer_resize(struct buffer *inst, size_t new_length)
+{   
+    // growing
+    if (inst->capacity < new_length) {
+        buffer_reserve(inst, new_length);
+    // shrinking
+    } else if (inst->capacity > new_length) {
+        for (size_t i = new_length; i < inst->capacity; ++i) {
+            ((destr)(inst->destructor))(inst->buffer[i]);
+        }
+        void *succeds = realloc(inst->buffer, new_length *sizeof(void*));
+        if (succeds) {
+            inst->buffer = succeds;
+        } else {
+            exit(0);
+        }
+    }
+    inst->capacity = new_length;
+    if (inst->length > new_length) {
+        inst->length = inst->capacity;
+    }
+}
+
+// Append to buffer (if no space, grow buffer by alloc_by items)
+void buffer_push_back(struct buffer *inst, void *item) 
+{   
+    size_t length = buffer_size(inst);
+    ++inst->length;
+    if (length < inst->capacity) {
+        void *new_item = buffer_at(inst, length);
+        if (item) {
+            memcpy(new_item, item, inst->item_size);
+        } else {
+            --inst->length;
+        }
+    } else {
+        buffer_reserve(inst, inst->alloc_by + length);
+    }
+}
+
+// Pointer to first item in buffer
+void *buffer_front(struct buffer *inst)
+{
+    if (inst) return inst->buffer[0];
+}
+
+// Pointer to last item in buffer
+void *buffer_back(struct buffer *inst)
+{
+    if (inst) return buffer_at(inst, buffer_size(inst) - 1);
+}
+
+// Take last item from buffer
+//  * return false if buffer is empty
+//  * copy item to "out" if not null
+//  * call destructor for item being removed
+//  * reduce length by 1
+//  * return true
+bool buffer_pop(struct buffer *inst, void *out)
+{
+    if (buffer_empty(inst)){
+        return false;
+    } else if (!buffer_back(inst)) {
+        return false;
+    } else {
+        memcpy(out, buffer_back(inst), inst->item_size);
+        --inst->length;
+        return true;
+    }
+}
+
+// Return true if buffer is empty (length == 0)
+bool buffer_empty(struct buffer *inst){
+    if (inst){
+       return (buffer_size(inst) == 0); 
+    } else {
+        exit(0);
+    }
+
+}
+
+// Number of items in the buffer (length)
+size_t buffer_size(struct buffer *inst)
+{
+    if (inst) {
+        return inst->length;
+    } else {
+        exit(0);
+    }
+}
+
+// Item size of the elements in the buffer(item_size)
+size_t item_size(struct buffer *inst)
+{
+    if (inst) {
+        return inst->item_size;
+    } else {
+        exit(0);
+    }
+}
+
+// Pointer to nth item in buffer
+void *buffer_at(struct buffer *inst, size_t index)
+{
+    if (index >= inst->length) {
+        return NULL;
+    }
+    return (inst->buffer[index]);
+}
+
+void destruct_size_t_item(void *item){
+    free(item);
+}
+
+
+
+// TESTING
+#if defined TEST_DYNAMIC_ARRAY
+
+struct buffer init()
```

Comment:
Here, we are returning a struct.  By the C standard, this can result in a lot of memory copying for larger structs (although some compilers will perform a "return value optimisation" to eliminate this copying).  It is better to pass a pointer to a struct as an argument, rather than copy-returning the struct.